velox/functions/prestosql/aggregates/benchmarks/ReduceAgg.cpp

```diff
@@ -0,0 +1,232 @@
+    {
+      auto sumResults = runOnce(makeGroupByPlan(kSum));
+      auto reduceAggResults = runOnce(makeGroupByPlan(kReduceAgg));
+
+      assertEqualResults({sumResults}, {reduceAggResults});
+    }
+  }
+
+  void runGlobalSum() {
+    runGlobal(kSum);
+  }
+
+  void runGroupBySum() {
+    runGlobal(kSum);
+  }
+
+  void runGlobalReduceAgg() {
+    runGlobal(kReduceAgg);
+  }
+
+  void runGroupByReduceAgg() {
+    runGroupBy(kReduceAgg);
+  }
+
+ private:
+  RowVectorPtr runOnce(const core::PlanFragment& planFragment) {
+    auto task = makeTask(planFragment);
+
+    std::vector<VectorPtr> results;
+    while (auto result = task->next()) {
+      results.push_back(result);
+    }
+
+    return copyResults(planFragment.planNode->outputType(), results);
+  }
+
+  RowVectorPtr copyResults(
+      const TypePtr& resultType,
+      const std::vector<VectorPtr>& results) {
+    if (results.empty()) {
+      return BaseVector::create<RowVector>(resultType, 0, pool());
+    }
+
+    auto totalCount = 0;
+    for (const auto& result : results) {
+      totalCount += result->size();
+    }
+
+    auto copy = BaseVector::create<RowVector>(resultType, totalCount, pool());
+    auto copyCount = 0;
+    for (const auto& result : results) {
+      copy->copy(result.get(), copyCount, 0, result->size());
+      copyCount += result->size();
+    }
+
+    return copy;
+  }
+
+  void runGlobal(const std::string& aggregate) {
+    folly::BenchmarkSuspender suspender;
+
+    auto plan = makeGlobalPlan(aggregate);
+    auto task = makeTask(plan);
+
+    suspender.dismiss();
+
+    vector_size_t numResultRows = 0;
+    while (auto result = task->next()) {
+      numResultRows += result->size();
+    }
+
+    folly::doNotOptimizeAway(numResultRows);
+  }
+
+  void runGroupBy(const std::string& aggregate) {
+    folly::BenchmarkSuspender suspender;
+
+    auto plan = makeGroupByPlan(aggregate);
+    auto task = makeTask(plan);
+
+    suspender.dismiss();
+
+    vector_size_t numResultRows = 0;
+    while (auto result = task->next()) {
+      numResultRows += result->size();
+    }
+
+    folly::doNotOptimizeAway(numResultRows);
+  }
+
+  core::PlanFragment makeGroupByPlan(const std::string& aggregate) {
+    return PlanBuilder()
+        .tableScan(inputType_)
+        .project({"i16 % 1234 as key", "i16"})
+        .partialAggregation({"key"}, {aggregate})
+        .finalAggregation()
+        .planFragment();
+  }
+
+  core::PlanFragment makeGlobalPlan(const std::string& aggregate) {
+    return PlanBuilder()
+        .tableScan(inputType_)
+        .partialAggregation({}, {aggregate})
+        .finalAggregation()
+        .planFragment();
+  }
+
+  std::shared_ptr<exec::Task> makeTask(core::PlanFragment plan) {
+    auto task = exec::Task::create(
+        "t",
+        std::move(plan),
+        0,
+        std::make_shared<core::QueryCtx>(executor_.get()));
+
+    task->addSplit("0", exec::Split(makeHiveConnectorSplit(filePath_->path)));
```

Comment:
Why do you write the data to a file and not just feed them using a Values operator since you already have them in memory?